src/librbd/mirror/snapshot/UnlinkPeerRequest.cc

```diff
@@ -91,15 +91,20 @@ void UnlinkPeerRequest<I>::unlink_peer() {
     finish(-EINVAL);
     return;
   }
-
-  // if there is or will be no more peers in the mirror snapshot and we have
-  // a more recent mirror snapshot, remove the older one
   if ((mirror_ns->mirror_peer_uuids.count(m_mirror_peer_uuid) == 0) ||
       (mirror_ns->mirror_peer_uuids.size() <= 1U && m_newer_mirror_snapshots)) {
-    m_image_ctx->image_lock.unlock_shared();
-    remove_snapshot();
-    return;
+
+    if (m_allow_remove) {
+      m_image_ctx->image_lock.unlock_shared();
+      remove_snapshot();
```

Comment:
Here's my thoughts:
>What is the point of dropping image_lock before calling remove_snapshot() when the first thing remove_snapshot() does is acquiring it again?

I see throughout `unlink_peer` whenever it exits, it makes sure to unlock before exiting. Perhaps that is a coding convention that was followed? I don't see a shared lock being held between functions anywhere in `UnlinkPeerRequest`.

>If image_lock wasn't dropped, could get_snap_info lookup and "failed to locate snapshot" error handler be dropped?

Agreed, if we did that we could remove the above.

>Why are the checks for whether the snapshot should actually be removed split between unlink_peer() and remove_snapshot() in the first place? The fact that remove_snapshot() can skip removal of snapshot causes nothing but confusion.

Agreed, perhaps we only enter `remove_snapshot()`, while still holding shared lock, if there is a newer snapshot and (peer list is empty OR last peer on list is m_mirror_peer_uuid).

Reply:
> I don't see a shared lock being held between functions anywhere in `UnlinkPeerRequest`.

This is because all other methods follow the `send` + `handle` state machine pattern, where `send` initiates some asynchronous work and registers `handle` as a callback for that.  The asynchronous work typically involves an OSD request going over the network or something else that would otherwise block, so releasing locks is pretty much a requirement.

The handover between `unlink_peer()` and `remove_snapshot()` is not like that, so releasing the lock really seems unmerited.

Reply:
I don't see/remember any particular reason why I decided to make `remove_snapshot()` to retake the lock, most likely just to follow the coding convention. It looks like ok to keep the lock and simplify the logic. Probably add some asserts at the beginning of `remove_snapshot()` that the lock is hold and snapshot exists.

Reply:
I removed `get_snap_info` from `remove_snapshot` and used values from `unlink_peer`. In this case, the shared lock isn't needed anymore before entering function. Therefore, I kept the `unlock_shared` contrary to what we discussed above.